SteamKit2/SteamKit2/Steam/Handlers/SteamMatchmaking/SteamMatchmaking.cs

```diff
@@ -0,0 +1,392 @@
+
+            foreach ( var lobby in lobbyList )
+            {
+                appLobbies[ lobby.SteamID ] = lobby;
+            }
+
+            Client.PostCallback( new LobbyListCallback(
+                body.app_id,
+                ( EResult )body.eresult,
+                lobbyList
+            ) );
+        }
+
+        void HandleLobbyData( IPacketMsg packetMsg )
+        {
+            var lobbyListResponse = new ClientMsgProtobuf<CMsgClientMMSLobbyData>( packetMsg );
+            CMsgClientMMSLobbyData body = lobbyListResponse.Body;
+
+            var lobbyId = new SteamID( body.steam_id_lobby );
+            var appLobbies = lobbies.GetOrAdd( body.app_id, k => new ConcurrentDictionary<SteamID, Lobby>() );
+            var existingLobby = appLobbies.ContainsKey( lobbyId ) ? appLobbies[ lobbyId ] : null;
+
+            List<Lobby.Member> memberList =
+                body.members.ConvertAll( member => new Lobby.Member(
+                    new SteamID( member.steam_id ),
+                    member.persona_name,
+                    Lobby.DecodeMetadata( member.metadata )
+                ) );
+
+            var updatedLobby = new Lobby(
+                lobbyId,
+                ( ELobbyType )body.lobby_type,
+                body.lobby_flags,
+                new SteamID( body.steam_id_owner ),
+                Lobby.DecodeMetadata( body.metadata ),
+                body.max_members,
+                body.num_members,
+                memberList,
+                existingLobby?.Distance,
+                existingLobby?.Weight
+            );
+
+            appLobbies[ lobbyId ] = updatedLobby;
+
+            Client.PostCallback( new LobbyDataCallback( body.app_id, updatedLobby ) );
+        }
+
+        void HandleUserJoinedLobby( IPacketMsg packetMsg )
+        {
+            var userJoinedLobby = new ClientMsgProtobuf<CMsgClientMMSUserJoinedLobby>( packetMsg );
+            CMsgClientMMSUserJoinedLobby body = userJoinedLobby.Body;
+
+            var lobbyId = new SteamID( body.steam_id_lobby );
+            var userId = new SteamID( body.steam_id_user );
+
+            var appLobbies = lobbies.GetOrAdd( body.app_id, k => new ConcurrentDictionary<SteamID, Lobby>() );
+            var lobby = appLobbies.ContainsKey( lobbyId ) ? appLobbies[ lobbyId ] : null;
+
+            if ( lobby == null )
+            {
+                // Unknown lobby
+                return;
+            }
+
+            int memberIndex = lobby.members.FindIndex( m => m.SteamID == userId );
+
+            if ( memberIndex >= 0 )
+            {
+                // Already in lobby
+                return;
+            }
+
+            var member = new Lobby.Member( new SteamID( body.steam_id_user ), body.persona_name );
+
+            lobby = lobby.Clone(); // Cloning to avoid mutating the original Lobby, as it was previously returned in a callback (not thread-safe).
+            lobby.members.Add( member );
+            appLobbies[ lobbyId ] = lobby;
```

Comment:
Potential race condition here between lines 323 and now.

Reply:
Hmm, aren't the handlers always called on the one thread (Client thread)?

Reply:
I'm not totally certain there's a potential race here. I think most HandleMsg based implementations expect to run in a single thread.

Reply:
They are, but I'd need to look closer at what the thread safety concerns _are_ here and how they're being addressed, since we have comments like `// ... (not thread-safe)`.